lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,683 @@
+              getModulePortName(parentModule, blockArg.getArgNumber())
+                  .getValue();
+          portWiring.push_back(std::move(wiring));
+          continue;
+        }
+
+        // Handle operations.
+        if (auto op = tappedOps.lookup(portAnno.anno)) {
+          // We currently require the target to be a wire.
+          // TODO: This should probably also allow wires?
+          auto wire = dyn_cast<WireOp>(op);
+          if (!wire) {
+            auto diag = blackBox.extModule.emitError(llvm::formatv(
+                "ReferenceDataTapKey on port {0} must be a wire", portName));
+            diag.attachNote(op->getLoc()) << "referenced operation is here:";
+            signalPassFailure();
+            continue;
+          }
+
+          // We currently require the target to be named.
+          // TODO: If we were to use proper cross-module reference ops in the IR
+          // then this could be anonymous, with ExportVerilog resolving the name
+          // at the last moment.
+          auto name = wire.nameAttr();
+          if (!name) {
+            auto diag =
+                wire.emitError("wire targeted by data tap must have a name");
+            diag.attachNote(blackBox.extModule->getLoc()) << llvm::formatv(
+                "used by ReferenceDataTapKey on port {0} here:", portName);
+            signalPassFailure();
+            continue;
+          }
+
+          wiring.prefices = instanceGraph.getAbsolutePaths(
+              wire->getParentOfType<FModuleOp>());
+          wiring.suffix = name.getValue();
+          portWiring.push_back(std::move(wiring));
+          continue;
+        }
+
+        // The annotation scattering must have placed this annotation on some
+        // target operation or block argument, which we should have picked up in
+        // the tapped args or ops maps.
+        signalPassFailure();
+        llvm::report_fatal_error(
+            llvm::formatv("ReferenceDataTapKey annotation was not scattered to "
+                          "an operation: {0}",
+                          portAnno.anno)
+                .str());
+      }
+
+      // Handle data taps on black boxes.
+      if (cls == strings.internalKeyClass) {
+        auto op = tappedOps.lookup(portAnno.anno);
+        if (!op) {
+          signalPassFailure();
+          llvm::report_fatal_error(
+              llvm::formatv(
+                  "DataTapModuleSignalKey annotation was not scattered to "
+                  "an operation: {0}",
+                  portAnno.anno)
+                  .str());
+        }
+
+        // Extract the internal path we're supposed to append.
+        auto internalPath = portAnno.anno.getAs<StringAttr>("internalPath");
+        if (!internalPath) {
+          blackBox.extModule.emitOpError(
+              llvm::formatv("DataTapModuleSignalKey annotation on port {0} "
+                            "missing \"internalPath\" attribute",
+                            portName));
```

Comment:
Instead of using format, I'd just use stream insertion:

blackBox.extModule.emitError("stuff") << portName << " other stuff";


Incidentally, emitOpError is a compiler hacker thing - it is a good thing to use for malformed IR etc, but you should use emitError for things users may encounter

Reply:
Nice, that makes sense. I've changed things around a bit to cater to this.